src/Web/Rails3/Session.hs

```diff
@@ -0,0 +1,156 @@
+
+    compareChecksum :: Digest SHA1 -> Either Text ByteString
+    compareChecksum checksum = if (computedChecksum == checksum) then (Right $ B64.decodeLenient b64) else (Left "[Rails3 Cookie] Checksum doesn't match")
+
+    computedChecksum :: Digest SHA1
+    computedChecksum = HMAC.hmacGetDigest (HMAC.hmac cookieSecret b64 :: HMAC SHA1)
+
+    (b64, hexChecksum) = let (a, b) = (breakSubstring delimiter $ urlDecode False x)
+                         in (a, BS.drop (BS.length delimiter) b)
+    delimiter = "--"
+
+    convertLeftToText e = case e of
+      Left l -> Left $ T.pack l
+      Right r -> Right r
+
+-- | Utility function to extract a named cookie from a 'Wai.Request'. In most
+-- Rails3 applications the cookie name is going to be of the format
+-- @_yourappname_session@. The exact name is /usually/ present in a file named
+-- @config\/initializers\/session_store.rb@
+extractEither :: CookieName -> Request -> Either T.Text ByteString
+extractEither (CookieName cname) req = maybeToEither "[Rails3 Cookie] No cookie header in the WAI request" (lookup "Cookie" (requestHeaders req))
+  >>= (return.parseCookies)
+  >>= (maybeToEither (T.concat ["[Rails3 Cookie] Cookie named '", decodeUtf8 cname, "' not found"])) . (lookup cname)
+
+-- NOTE: Please refer to
+-- http://blog.bigbinary.com/2013/03/19/cookies-on-rails.html to understand how
+-- a Rails3 cookie is encoded (NOT encyrpted). Encryption of session cookies
+-- only began in Rails4. Rails3 marshals a RubyObject and base64 encodes it to
+-- store it as a cookie. To ensure that it cannot be tamped with, it also adds
+-- an HMAC computed with the help of a secret key/value/token.
+
+-- | Utility function to extract a cookie called @CookieName@ from the
+-- 'Wai.Request', decode it using the supplied @Secret@
+extractAndDecodeEither :: (CookieName, Secret) -> Request -> Either T.Text RubyObject
+extractAndDecodeEither (cookieName, cookieSecret) req = (extractEither cookieName req) >>= ((decodeEither cookieSecret) . Cookie)
+
+safeHead :: [a] -> Maybe a
+safeHead []    = Nothing
+safeHead (x:_) = Just x
+
+lookupKey :: (Rubyable a) => (BS.ByteString, RubyStringEncoding) -> RubyObject -> Maybe a
+lookupKey key robj = (fromRuby robj :: Maybe (Map.Map (BS.ByteString, RubyStringEncoding) RubyObject))
+  >>= Map.lookup key
+  >>= fromRuby
+
+-- | Lookup the Warden\/Devise UserIds from a decoded cookie. __Please note,__ a
+-- cookie may contain multiple UserIds, because it /seems/ that it is possible
+-- to be logged-in as multiple users simultaneously, if you define [multiple
+-- user
+-- models](https://github.com/plataformatec/devise/wiki/How-to-Setup-Multiple-Devise-User-Models)
+-- (the underlying data-structure allows it, as well).
+lookupUserIds :: (Num a) => RubyObject -> Maybe (NonEmpty a)
+lookupUserIds robj =
+  lookupKey ("warden.user.user.key", UTF_8) robj
+  >>= (\x -> fromRuby x :: Maybe [RubyObject]) -- [[int, int int], "random string"]
```

Comment:
Wouldn't this be cleaner in do notation?

Reply:
Will make the change.

Reply:
Unable to make this change sensibly. Giving each intermediate value a sensible name is harder than I thought. If it's not too much of trouble, can we let this be as is?